plutus-core/plutus-core/test/Evaluation/Spec.hs

```diff
@@ -1,11 +1,247 @@
+      -- TODO: remove this
+      List.\\ [ VerifySignature,
+                VerifyEcdsaSecp256k1Signature,
+                VerifySchnorrSecp256k1Signature
+              ]
+
+prop_builtinsDon'tThrow :: DefaultFun -> Property
+prop_builtinsDon'tThrow bn = property $ do
+  (args, argStrings) <- unzip <$> forAllNoShow (genArgs bn)
+  let (res, logs) = runEmitter . runExceptT $ eval args argStrings
+  case res of
+    Right _ -> success
+    Left err -> do
+      liftIO $ do
+        putStrLn "Builtin function evaluation failed"
+        putStrLn $ "Function: " <> show bn
+        putStrLn $ "Arguments: " <> show argStrings
+        putStrLn $
+          "Error: "
+            <> ( case err of
+                   KnownTypeEvaluationFailure -> "KnownTypeEvaluationFailure"
+                   KnownTypeUnliftingError e ->
+                     "KnownTypeUnliftingError: "
+                       <> show (pretty e)
+               )
+        putStrLn $ "Execution log: " <> show logs
+      failure
+  where
+    meaning :: BuiltinMeaning Term (CostingPart DefaultUni DefaultFun)
+    meaning = toBuiltinMeaning bn
+
+    eval :: [Term] -> [String] -> MakeKnownM Term
+    eval args0 argStrings = case meaning of
+      BuiltinMeaning _ _ runtime -> go (_broRuntimeScheme runtime) (_broImmediateF runtime) args0
+      where
+        go ::
+          forall n.
+          RuntimeScheme n ->
+          ToRuntimeDenotationType Term n ->
+          [Term] ->
+          MakeKnownM Term
+        go sch f args = case (sch, args) of
+          (RuntimeSchemeArrow sch', a : as) -> do
+            res <- liftEither (f a)
+            go sch' res as
+          (RuntimeSchemeResult, []) -> f
+          (RuntimeSchemeAll sch', _) -> go sch' f args
+          _ -> error $ "Wrong number of args for builtin " <> show bn <> ": " <> show argStrings
+
+-- | Generate arguments to a builtin function
+genArgs :: DefaultFun -> Gen [(Term, String)]
+genArgs bn = case bn of
+  -- These functions are partial, so we manually generate their arguments.
+  IndexByteString -> do
+    s <- Gen.utf8 (Range.linear 5 100) Gen.enumBounded
+    i :: Integer <- fromIntegral <$> Gen.int (Range.linear 0 (BS.length s - 1))
+    pure [mkArg s, mkArg i]
+  HeadList -> pure . pure . mkArg =<< Gen.list (Range.linear 1 100) genInteger
+  TailList -> genArgs HeadList
+  UnConstrData -> pure . pure . mkArg =<< genConstr 5
+  UnMapData -> pure . pure . mkArg =<< genMap 5
+  UnListData -> pure . pure . mkArg =<< genList 5
+  UnIData -> pure . pure . mkArg =<< genI
+  UnBData -> pure . pure . mkArg =<< genB
+  -- The rest are total functions and we generate the arguments based on their `TypeScheme`s.
+  _ -> sequenceA $ case meaning of
+    BuiltinMeaning tySch _ _ -> go tySch
+      where
+        go :: forall args res. TypeScheme Term args res -> [Gen (Term, String)]
+        go = \case
+          TypeSchemeResult    -> []
+          TypeSchemeArrow sch -> genArg (typeRep @(Head args)) : go sch
```

Comment:
This is perfectly fine, but we have a convenience function for this defined in `PlutusCore.Builtin.TypeScheme`:

```
argProxy :: TypeScheme val (arg ': args) res -> Proxy arg
```

Although I'm not really sure if using it is more convenient than just calling `Head`.

Reply:
Yeah, since `typeRep` doesn't expect a `Proxy`, using `argProxy` would be a bit awkward.